### ISSUE TITLE
Floating point errors in `.align_to()`

### ISSUE THREAD

**Issue by connorferster** [robbievanleeuwen/section-properties#164]
**Describe the bug**
The below code:
```python
import sectionproperties.pre.library.bridge_sections as bridge_sections
import sectionproperties.pre.library.primitive_sections as standard_sections
from sectionproperties.pre.pre import Material
from sectionproperties.analysis.section import Section
from sectionproperties.pre.geometry import CompoundGeometry

Dslab, w, t_f = 180, 2100, 75

precast = Material(
    name="65 MPa",
    elastic_modulus=37.4e3,
    poissons_ratio=0.2,
    yield_strength=65,
    density=2.4e-6,
    color="grey",
)
insitu = Material(
    name="40 MPa",
    elastic_modulus=32.8e3,
    poissons_ratio=0.2,
    yield_strength=40,
    density=2.4e-6,
    color="lightgrey",
)

super_t = bridge_sections.super_t_girder_section(girder_type=5, w=w, material=precast)
slab = standard_sections.rectangular_section(d=Dslab, b=w, material=insitu).align_center(super_t).align_to(super_t, "top")

geom = CompoundGeometry([super_t, slab])
geom.plot_geometry()
geom
``` 
Produces this...

![image](https://user-images.githubusercontent.com/38441489/147904830-6859a105-3219-4cb7-a94b-7bbeb03a1ea3.png)

But it SHOULD produce this...(Note the red svg denoting "invalid" geometry and the hole indicated on the plot)

![image](https://user-images.githubusercontent.com/38441489/147904879-ef171f5c-715b-4400-bd59-574cfb1e7360.png)

**Additional context**
This seems to be a problem with shapely. I opened an [issue with shapely](https://github.com/shapely/shapely/issues/1279) after realizing that one of their examples in their docs does not reproduce as given (another example of the same problem): https://shapely.readthedocs.io/en/stable/manual.html#polygons

```python
exter = [(0, 0), (0, 2), (2, 2), (2, 0), (0, 0)]
inter = [(1, 0), (0, 1), (0.5, 1.5), (1.5, 0.5), (1, 0)]
(Polygon(exter) - Polygon(inter)) # Creates valid MultiPolygon even though two points in common
```

Produces this...
![image](https://user-images.githubusercontent.com/38441489/147905075-3c089fcd-843b-4919-8f96-8c156cd4de54.png)

**Something else that's funny (odd, not 'ha ha')**

If I take the `geom` object that was created in the bridge section code (at top) and break it out with `.points` as such:

```python
geom = CompoundGeometry(MultiPolygon([Polygon(slab.points), Polygon(super_t.points)]))

# Instead of...

geom = CompoundGeometry([slab, super_t])
```

Then I also get this...

![image](https://user-images.githubusercontent.com/38441489/147905296-1bd29afb-63a2-4228-a49f-b0832f3b7ab1.png)

I don't fully understand but I think it is shapely related... Will post more on this as I learn more.

**Comment by connorferster** [comment]
A more robust algorithm for `Geometry.align_to()` might be to instead replace source ordinates with the target ordinates instead of calculating offsets with them.

e.g. If aligning Geometry A to the top of Geometry B, the y-ordinates matching the minimum y-extent of Geometry A would be changed so that they are the same as the maximum y-extent of Geometry B. This way, they remain exact and no rounding error occurs.

**Comment by connorferster** [comment]
There is also this:
https://gis.stackexchange.com/questions/188622/rounding-all-coordinates-in-shapely/188739#188739

**Comment by robbievanleeuwen** [comment]
Hi @connorferster, I agree that the precision method seems to be a good approach, i.e. ``offset = round(align_to_coord - self_align_coord, precision)``.

How do you see the stackexhange ``_set_precision()`` method working? Every time the coordinates are modified this method is called? Perhaps that's a good idea? Maybe a note to the user in the docs as well that there is a defined minimum precision e.g. 8 and that units should be chosen to avoid precision rounding issues, i.e. don't model your small section in kilometers.

Reminds of the issues early kerbal space program had with precision!

**Comment by normanrichardson** [comment]
@connorferster I'm not sure if I'm 100% following the issue, but the hole is not being detected, seems like [here](https://github.com/robbievanleeuwen/section-properties/blob/f935aa43e0fd282359ec0571ac85a1073b6e3443/sectionproperties/pre/geometry.py#L1697-L1707) is where it should happen. I can see that there are two sets of logic for MultiPolygon and Polygons. I can see why this is done: to catch holes interior to each Polygon. It seems to miss holes that are formed by the two polygons' exteriors though.

I would like to propose a new logic that I think will work for both MultiPolygon and Polygon. I will use your example to demonstrate the logic:

```python
from shapely import geometry
from shapely.geometry.point import Point
import sectionproperties.pre.library.bridge_sections as bridge_sections
import sectionproperties.pre.library.primitive_sections as standard_sections
from sectionproperties.pre.pre import Material
from sectionproperties.analysis.section import Section
from sectionproperties.pre.geometry import CompoundGeometry, Geometry
from shapely.geometry import box, Point

Dslab, w, t_f = 180, 2100, 75

precast = Material(
    name="65 MPa",
    elastic_modulus=37.4e3,
    poissons_ratio=0.2,
    yield_strength=65,
    density=2.4e-6,
    color="grey",
)
insitu = Material(
    name="40 MPa",
    elastic_modulus=32.8e3,
    poissons_ratio=0.2,
    yield_strength=40,
    density=2.4e-6,
    color="lightgrey",
)

super_t = bridge_sections.super_t_girder_section(girder_type=5, w=w, material=precast)
slab = standard_sections.rectangular_section(d=Dslab, b=w, material=insitu).align_center(super_t).align_to(super_t, "top")

geom = CompoundGeometry([super_t, slab])

# a proposed fix to hole detection in CompoundGeometry
b_box = box(*geom.geom.bounds)
sub = b_box - geom.geom
hole_polys = [ii for ii in sub if not b_box.exterior.intersects(ii)]
holes = [hole.representative_point() for hole in hole_polys]
[print(hole) for hole in holes]
```

Essentially, it takes a bounding box and cuts out the geometry, then there is some filtering to find out which cutouts are interior to the section. Let me know your thoughts, there may be some issues I'm not seeing.

**Comment by normanrichardson** [comment]
"It seems to miss holes that are formed by the two polygons' exteriors though"
Correction: it seems unary_union is not working as expected in this case. For this case we should expect a merged Polygon... and it is producing a Multipolygon, and that is why the hole formed by the two exteriors is not being picked up.

**Comment by normanrichardson** [comment]
Unfortunately, I'm now beginning to see the same issues with this method:

```python
geom = CompoundGeometry([super_t, slab])
geom = geom.rotate_section(30)

# try proposed fix to hole detection in CompoundGeometry
b_box = box(*geom.geom.bounds)
sub = b_box - geom.geom
hole_polys = [ii for ii in sub if not b_box.exterior.intersects(ii)]
holes = [hole.representative_point() for hole in hole_polys]
[print(hole) for hole in holes]
```

Scratch that idea.

**Comment by connorferster** [comment]
@robbievanleeuwen 

So, it seems that just rounding the offset some decimal places does not work: it seems like it can create a small overlap between two geometries which causes triangle to spiral out of control. 

I think that `numpy.clip` could hold a solution but it will be a bit tricky to implement. Started playing with it this evening. Will take some thinking.

**Comment by connorferster** [comment]
Scratch that...numpy.clip can only handle floating point errors in one direction for each of `min` and `max`. Back the drawing board.

**Comment by connorferster** [comment]
Gah! Simple solution: round the resulting coordinates, not the offset. <eyeroll> 

In `Geometry.compile_geometry()`, which generates the `.points`, `.facets`, etc. the coordinates are already being rounded to `self.tol` which is set to `12` decimal places. This is why the geometry can be created without holes but when using the generated points in a new geometry with `.from_points()`, we get what we want. 

Totally forgot about that tolerance thing in the creation of `.points`. Going to re-order `Geometry.__init__()` so that the shapely object is exactly the same as the generated points because, right now, there is a a slight difference between the geometry you see and the points that are generated from it.

**Comment by connorferster** [comment]
Fixed in #167
